server/sonar-server/src/test/java/org/sonar/server/user/ServerUserSessionTest.java

```diff
@@ -80,6 +80,15 @@ public void setUp() throws Exception {
   }
 
   @Test
+  public void anonymous_is_not_logged_in_and_does_not_have_login() throws Exception {
```

Comment:
I don't see any test of the caching mechanism, except for `hasSystemAdminister` and `getGroups`. From statically looking at the code, this is a miss (see my other comment).